### ISSUE TITLE
'Cheating' Split spells

### ISSUE THREAD

**Issue by ShawnieBoy** [magarena/magarena#668]
The main obstacle with split spells is that they have two values everywhere except on the stack. This is generally in regards to color and casting cost.

If we can't have the two halves properly represented, then maybe we could treat them as two separate cards.

These card-pairs would need to move in unison, so if one half is searched for, the other half comes as well. The same in regards to discarding. These are the main barriers with regards to their properties e.g Discard a black card, Search for card with converted mana cost 2 or less.

One side-effect of this may mean that the two halves would technically be in separate places within the deck. But when one half moves, the other moves as well. As long as one half isn't counted in card totals (while in the hand or deck for example) there would be no obvious changes.

**Comment by ShawnieBoy** [comment]
Just looking at Split Cards again...

Just a question really, hopefully @melvinzhang can answer - Why can't there be multiple HandCastActivations on a card? Seems to be what Split needs.

**Comment by melvinzhang** [comment]
@ShawnieBoy Why do you think you can't have multiple? MagicAbility.AlternateCost is implemented as adding a new MagicHandCastActivation.

**Comment by ShawnieBoy** [comment]
It's in MagicCardDefintion.setHandAct - but I guess that kind of makes sense, as each each half would have it's own CardDefinition, but the MagicCard would have both activations available (With the name of the half instead of "Cast")

**Comment by melvinzhang** [comment]
setHandAct is used by Affinity only, all other hand activations use addHandAct.

**Comment by melvinzhang** [comment]
As Affinity replaces the default casting activation rather than add a new one.

**Comment by ShawnieBoy** [comment]
Ahh, I'm with you - yes, that makes sense, setting the HandActivation to a new value.

**Comment by melvinzhang** [comment]
My bad, Affinity (for scripted cards) + bunch of cards in groovy code (Stone Idol Trap, Draco, etc).

**Comment by ShawnieBoy** [comment]
I remember using it now for the 'Avatar of' cards

**Comment by ShawnieBoy** [comment]
Split Cards are one of those things which are 'so close' to being implemented. I think it's just the logic for determining the converted costs of the card which will still need work. As we're unable to return multiple values, and thankfully most of the mtg rulings don't need the actual values, it's just a 'Yes' or 'No' boolean answer.

Does this card have a converted casting cost of X? Yes/No

**Comment by melvinzhang** [comment]
Not to forget 'more than', 'more than or equal', 'less than', 'less than or equal'. Seems to be only a small finite set, perhaps we can add methods for each one.

**Comment by ShawnieBoy** [comment]
Yes - of course. But it's the only change that would have a larger impact on how we determine converted costs than the check of both sides for color.

**Comment by melvinzhang** [comment]
Some things the sum though, eg 

"Infernal Genesis has an ability that reads, “At the beginning of each player’s upkeep, that player puts the top card from his or her library into his or her graveyard. He or she then puts X 1/1 black Minion creature tokens onto the battlefield, where X is that card’s converted mana cost.” If the top card of your library is Assault/Battery, which has converted mana costs of 1 and 4, when this ability resolves, you simultaneously put one and four creature tokens onto the battlefield, for a total of five."
--- http://magiccards.info/rule/708-split-cards.html

**Comment by ShawnieBoy** [comment]
Yes, that would use a simple getConvertedCost() - but determining what the converted cost of a card is - "Discard a card with converted mana cost of 2 or less", can't use getConvertedCost() to check.

**Comment by ShawnieBoy** [comment]
708.6c will be the ultimate test :grinning:

**Comment by ShawnieBoy** [comment]
Will also be useful to look at for the changes to double-faced cards. A transformed card will use the 'face-down' value for it's converted cost, except when it's copied (as it has no face-down half)

**Comment by melvinzhang** [comment]
@ShawnieBoy create separate issues please, keep this one for split cards only.

**Comment by ShawnieBoy** [comment]
:+1: Will do

**Comment by melvinzhang** [comment]
With the rules change to CMC of split cards, we can now implement split cards in a decent way without 'cheating'

**Comment by melvinzhang** [comment]
Would it make things simpler to have a card def for the combined card that references card defs for each half? Sort of like a triple faced card. 

This would have one card named "Fire and Ice" which has the combined characteristics and it references two other card def one for "Fire" and one for "Ice". It gets put on the stack similar to Morph spells, so when "Fire" is cast, the source card is "Fire and Ice" but the characteristics as the ones for "Fire", similar to how the characteristics of a Morph spell on the stack is the hidden Morph card def.

**Comment by ShawnieBoy** [comment]
I was thinking about that myself. The issue I thought of was being able to cast a spell without the combined cost overriding what you are able to cast from the hand.

That could be solved by having two casting abilities on the combined card, somewhat like a modular card.

**Comment by ShawnieBoy** [comment]
Currently split spells are defined as a dominant 'first' half and a second-half, much like transform. This would still work, as long as the hasType, hasColor etc of the card also returns the other half. Similar to how a transformed card uses the converted cost of the other side.
The card would then only require an additional card activation ability for the other half - or this could be pulled automatically from the other cardDef.

**Comment by Fulmene** [comment]
I'm currently working on this [here](https://github.com/Fulmene/magarena/tree/split-card). After some initial testing with Status // Statue, some concerns I have are:

- Showing the card name. Currently it's shown as "Status // Statue (Status)" in deckbuilder and just Status in game.
- Interaction with cards that let it be cast from the graveyard like Snapcaster Mage. Currently with Status // Statue I can only cast Status from the graveyard.

**Comment by melvinzhang** [comment]
> just Status in game

That is because "Status" is the main half, the main half second half implementation would have this problem. It is probably better to have a "Split version" and two separate halves. The split version would have the name "Status // Statue" and have a mana cost that is the sum of the two halves so that we can implement this rule: "In every zone except the stack, the characteristics of a split card are those of its two halves combined."

**Comment by Fulmene** [comment]
> It is probably better to have a "Split version" and two separate halves.

This is what I'm thinking. It would be easier to implement the main card as a separate entity from both halves. Something like this?

Status____Statue.txt
```
name=Status // Statue
image=(URL here)
split=Status,Statue
```

The card type, cost, etc should be derived from the halves and not specified on the split version itself.

**Comment by melvinzhang** [comment]
Yes, that is exactly what I had in mind.
